### ISSUE TITLE
HopSkipJump Attack returns 'nan'-Values and enters Infinite Loop

### ISSUE THREAD

**Issue by FlorianMerkle** [bethgelab/foolbox#670]
I am not sure if the HopSkipJump attack is maintained, as it is not documented. However, it is part of the official API so I thought I report the problem anyways.

We realized that sometimes the attack returns 'nan'-tensors and sometimes enters an infinite loop.

The nan return values appear when the model is naturally misclassifying one of the given samples in the batch. I tried to track down the problem and found this line to be problematic:
https://github.com/bethgelab/foolbox/blob/43858a0bfc2b83023ae87954f738194708f8d923/foolbox/attacks/hop_skip_jump.py#L254

It seems that in that specific case both, divisor and denominator, are zero, causing the nan-values. This can happen when delta becomes zero (which it will after the first step when the distance is zero).
https://github.com/bethgelab/foolbox/blob/43858a0bfc2b83023ae87954f738194708f8d923/foolbox/attacks/hop_skip_jump.py#L364

We were able to fix this and get sensible results by adding a small constant to delta in line 254.

About the infinite loop here:
https://github.com/bethgelab/foolbox/blob/43858a0bfc2b83023ae87954f738194708f8d923/foolbox/attacks/hop_skip_jump.py#L166-L174

Interestingly, torch models might return real values when inputting nan-tensors, the first convolutional layer of our model returned -inf values, which in turn become zero when passed into a ReLU function and the model will return logits. The attack will not enter an infinite loop as long none of the samples in the batch are actually the class, the model predicts for full nan-tensors.

**Comment by zimmerrol** [comment]
Thanks for reporting this! I just merged a PR adressing this issue; the fix will be available with the next update for foolbox.
